mozjs/src/gc/root.rs

```diff
@@ -19,60 +20,64 @@ use mozjs_sys::jsgc::ValueArray;
     feature = "crown",
     crown::unrooted_must_root_lint::allow_unrooted_interior
 )]
-pub struct RootedGuard<'a, T: 'a + RootKind + Initialize> {
+pub struct RootedGuard<'a, T: 'a + RootKind> {
     root: &'a mut Rooted<T>,
 }
 
-impl<'a, T: 'a + RootKind + Initialize> RootedGuard<'a, T> {
+impl<'a, T: 'a + RootKind> RootedGuard<'a, T> {
     pub fn new(cx: *mut JSContext, root: &'a mut Rooted<T>, initial: T) -> Self {
-        root.ptr = initial;
+        root.ptr.write(initial);
         unsafe {
             root.add_to_root_stack(cx);
         }
         RootedGuard { root }
     }
 
     pub fn handle(&'a self) -> Handle<'a, T> {
-        Handle::new(&self.root.ptr)
+        Handle::new(&self)
     }
 
     pub fn handle_mut(&mut self) -> MutableHandle<T> {
-        unsafe { MutableHandle::from_marked_location(&mut self.root.ptr) }
+        unsafe { MutableHandle::from_marked_location(self.deref_mut()) }
     }
 
     pub fn get(&self) -> T
     where
         T: Copy,
     {
-        self.root.ptr
+        // SAFETY: The rooted value is initialized as long as we exist
+        unsafe { self.root.ptr.assume_init() }
     }
 
     pub fn set(&mut self, v: T) {
-        self.root.ptr = v;
+        self.root.ptr.write(v);
```

Comment:
We will leak the previous initialized value if this is invoked.

Reply:
Nice catch!